### PR TITLE
Better handling for deleted apps

### LINKED ISSUE #226: Deleting an application in shinyapps.io causes a 404 when re-deploying
Once an application has been deployed, the dcf files stores the application id which is used for subsequent interactions. If the application is deleted via browser interaction, further rsconnect calls result in a 404 error because the application is no longer server side. The current work around is to manually delete the dcf file, usually after [having had to ask](https://groups.google.com/d/msg/shinyapps-users/tFCUEBj8hzg/sFZOmwloAwAJ) for [help in a forum](https://groups.google.com/d/msg/shinyapps-users/R6NMc9Ymf5E/IbTXjwlxBAAJ).

Consider handling 404 errors explicitly with a more informative message or actually removing the dcf file itself.

### PR DESCRIPTION
**Pull request by hadley**
If `deployApp()` determines that it's publishing to an app that doesn't exist, it offers the user the choice to create a new app. Fixes #226.

Includes somewhat of a revamp to `syncAppMetadata()` to delete deployments that don't correspond to known applications. Interactive testing of this function revealed some other problems that I fixed, along with generally tidying up the code.

* [x] Add some tests for `syncAppMetadata()`
* [x] Add tests for `applicationDeleted()`

### CODE REVIEW

File: R/applications.R [hunk 1]
```diff
@@ -255,77 +254,66 @@ showLogs <- function(appPath = getwd(), appFile = NULL, appName = NULL,
...
     # don't sync if within the configured time window
-    if (as.numeric(lastSyncTime) + syncWindow > now) {
+    if (as.numeric(lastSyncTime) + cacheHours * 3600 > now) {
       next
     }
-
-    # don't sync non-connect apps
-    if (!isConnectInfo(server = deploys[i, "hostUrl"])) {
+    # or if published to RPubs
+    if (!isRPubs(curDeploy$server)) {
```
Comment (hadley) [inline R/applications.R:289]: I tightened this condition because `getApplication()` works for shinyapps.

File: R/applications.R [hunk 2]
```diff
@@ -255,77 +254,66 @@ showLogs <- function(appPath = getwd(), appFile = NULL, appName = NULL,
...
-#' Update the metadata for requested application across all deployments
+#' Update the deployment records for applications published to Posit Connect.
+#' Updates application title and URL, and deletes records for deployments
+#' where the application has been deleted on the server.
 #'
 #' @param appPath The path to the directory or file that was deployed.
-#'
-#' @note This function does not update metadata for Shiny and rpubs apps
-#'
+#' @param cacheHours Number of hours to wait between confirming
```
Comment (hadley) [inline R/applications.R:264]: Adding this argument made it easier to test. But given how little it takes to run and that `rsconnect.metadata.sync.hours` doesn't appear to be used anywhere else, I think it makes sense to remove the `lastSyncTime` component all together.

File: R/deployApp.R [hunk 3]
```diff
@@ -474,6 +484,34 @@ runDeploymentHook <- function(appDir, option, verbose = FALSE) {
   hook(appDir)
 }
 
+applicationDeleted <- function(client, target, recordPath) {
+  header <- "Failed to find existing application on server; it's probably been deleted."
+  not_interactive <- "Use {.fn forgetDeployment} to remove outdated record and try again."
```
Comment (aronatkins) [inline R/deployApp.R:489]: We could also suggest `applications()` to see other deployments.

File: R/applications.R (hunk 2 above)
Comment (aronatkins) [reply R/applications.R:264]: I don't know of any calls from the IDE to this function; I feel like we could remove all notion of "time" (no skipped updates, no update timestamps).

**Review by aronatkins — APPROVED** [review]

**Pull request merged**